parser.y

```diff
@@ -396,6 +398,18 @@ decl: declaration_specifiers init_declarator_opt SEMICOLON {
       $$ = std::move(init_decl);
     } else {
       auto decl = std::move(std::get<std::unique_ptr<DeclNode>>(decl_specifers));
+      auto rec_decl = dynamic_cast<RecordDeclNode*>(decl.get());
+      assert(rec_decl);
+      if (auto* rec_var_decl = dynamic_cast<RecordVarDeclNode*>(init_decl.get())) {
+        // A struct or union variable.
+        rec_var_decl->type = ResolveType(std::move(rec_decl->type), std::move(rec_var_decl->type));
+        decl = std::move(init_decl);
+      } else if (auto* arr_decl = dynamic_cast<ArrDeclNode*>(init_decl.get())) {
+        // An array with struct or union elements.
+        arr_decl->type = ResolveType(std::move(rec_decl->type), std::move(arr_decl->type));
+        decl = std::move(init_decl);
+      }
+
```

Comment:
This code can be simplified by utilizing the inheritance hierarchy. Since the data member 'type' is held by the base class `DeclNode`, there is no need to cast to the concrete classes. 
```suggestion
      if (init_decl) {
        init_decl->type = ResolveType(std::move(rec_decl->type), std::move(init_decl->type));
        decl = std::move(init_decl);
      }
```
> [!IMPORTANT]
> Please notice the check `if (init_decl)`, since we are having unnamed struct/union in our test cases, the `init_declarator_opt` could be a null pointer. I kindly request that you solve the unnamed primitive types as also, which is included by `/* TODO: If the init declarator doesn't present, e.g., ``int;``, the declaration is still valid. */`, so that the `if (init_decl)` check can be moved one level above to cover both primitive and user defined types.